### PR TITLE
WIP Metaflow on Airflow 

### PR DESCRIPTION
**Pull request by valayDave**
## Current Support
- Datastore: `s3`
- Metadata: `service`
- compute: `@kubernetes`
- Internal decorators support : `@catch`,`@retry`, `@timeout`, `@environment`, `@schedule`, `@conda`, `@card`
- Parameters 

## Limitations
**No support for Foreachs (Coming In subsequent commits.)**

## Code Status :
- Unpolished. 
- Lots of TODOs. 

## Massive TODOs: 
- [x] Linear Workflow
- [x] Setting Identifiers correctly (task-id / run-id)
- [x] Configurable Parameters
- [x] Support for Other decorators like  `@catch`,`@retry`, `@timeout`, `@environment`,  `@conda`, `@card`
- [x] Branching 
- [x] Support for Max workers. 
- [x] Credential Management. 
- [ ] `IncludeFile` Testing 
- [x] Test with `@schedule` properly. 
- [x] Support `@project` decorator
- [x] K8s Secrets
- [x] Worker pools
- [ ] Foreach
- [ ] Test Cases
- [ ] Alert Decorator

### CODE REVIEW

File: metaflow/plugins/airflow/airflow_utils.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE will work for linear/branched workflows.
+# ti.task_id is the stepname in metaflow code.
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE uses a jinja filter called `task_id_creator` which helps
+# concatenate the string using a `/`. Since run-id will keep changing and stepname will be
+# the same task id will change. Since airflow doesn't encourage dynamic rewriting of dags
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = "arf-{{ [run_id, ti.task_id ] | task_id_creator  }}"
```
Comment (valayDave) [inline metaflow/plugins/airflow/airflow_utils.py:30]: This is the macro we use to set `Task.id`

Comment (valayDave) [reply metaflow/plugins/airflow/airflow_utils.py:30]: reasoning for it's choice is given in the code comments.

File: metaflow/plugins/airflow/airflow_utils.py [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+    # non-alphahanumeric, replace it with 'u' to make it valid
+    sanitized_val = re.sub("^[^A-Z0-9a-z]", "u", re.sub(r"[^A-Za-z0-9.\-_]", "_", val))
+    return sanitized_val[:57] + "-" + hash[:5]
+
+
+def hasher(my_value):
+    return hashlib.md5(my_value.encode("utf-8")).hexdigest()
+
+
+def task_id_creator(lst):
```
Comment (valayDave) [inline metaflow/plugins/airflow/airflow_utils.py:56]: This is the templating function that helps create task-id from run-id and stepname.

Comment (valayDave) [reply metaflow/plugins/airflow/airflow_utils.py:56]: This helps ensure that taskids dont change during retry attempts.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+        # todo : resolve schedule interval based on decorator.
+        self.schedule_interval = self._get_schedule()
+        self._file_path = file_path
+
+    def _get_schedule(self):
+        schedule = self.flow._flow_decorators.get("schedule")
+        if schedule:
+            return schedule.schedule
+        # Airflow requires a scheduling arguement so keeping this
+        return "*/2 * * * *"
```
Comment (valayDave) [inline metaflow/plugins/airflow/airflow_compiler.py]: TODO : What should be defaults.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+
+
+class Airflow(object):
+
+    task_id = AIRFLOW_TASK_ID_TEMPLATE_VALUE
+    task_id_arg = "--task-id %s" % task_id
+
+    # Airflow run_ids are of the form : "manual__2022-03-15T01:26:41.186781+00:00"
+    # Such run-ids break the `metaflow.util.decompress_list`; this is why we hash the runid
+    run_id = "%s-$(echo -n {{ run_id }} | md5sum | awk '{print $1}')" % AIRFLOW_PREFIX
```
Comment (valayDave) [inline metaflow/plugins/airflow/airflow_compiler.py]: We try to retrieve md5-hash of run-id over here to fit it for airflow

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+                # Todo : Handle This case
+                pass
+
+            if any_incoming_node_is_foreach:
+                # todo : Handle split index for for-each.
+                # step.append("--split-index $METAFLOW_SPLIT_INDEX")
+                pass
+
+        # ! HACK : Remove Below Line; Only here temporarily.
+        env.update({k: v for k, v in os.environ.items() if "AWS" in k})
```
Comment (valayDave) [inline metaflow/plugins/airflow/airflow_compiler.py]: HACK HACK HACK. This is for very temporary testing purposes. Will be removed

File: metaflow/plugins/airflow/compute/k8s.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        retry_delay=dict(seconds=retry_delay.total_seconds()) if retry_delay else None,
+        retries=retries,
+        env_vars=[dict(name=k, value=v) for k, v in env.items()],
+        labels=labels,
+        is_delete_operator_pod=True,
+    )
+    return k8s_operator_args
+
+
+# todo : below code block only for reference. Remove it later.
```
Comment (valayDave) [inline metaflow/plugins/airflow/compute/k8s.py]: The entire code block below this will be removed. It is only here for checking argument types.

File: metaflow/datastore/task_datastore.py [hunk 7]
```diff
@@ -264,36 +296,7 @@ def pickle_iter():
                     if isinstance(force_v4, bool)
                     else force_v4.get(name, False)
                 )
-                if do_v4:
```
Comment (savingoyal) [inline metaflow/datastore/task_datastore.py]: This change doesn't seem to be doing anything specific here. Can we drop it?

File: metaflow/metaflow_config.py [hunk 8]
```diff
@@ -197,6 +197,11 @@ def from_conf(name, default=None):
 )
 #
 
+###
+# Airflow Configuration
+###
+AIRFLOW_STATE_MACHINE_PREFIX = from_conf("METAFLOW_AIRFLOW_STATE_MACHINE_PREFIX")
```
Comment (savingoyal) [inline metaflow/metaflow_config.py]: `METAFLOW_AIRFLOW_STATE_MACHINE_PREFIX ` - I don't think this environment variable is needed. Let's drop it.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+from metaflow import current
+from metaflow.exception import MetaflowException
+import re
+
+VALID_NAME = re.compile("[^a-zA-Z0-9_\-\.]")
+
+
+def resolve_state_machine_name(name):
+    def attach_prefix(name):
+        if AIRFLOW_STATE_MACHINE_PREFIX is not None:
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: There is no need for `STATE_MACHINE_PREFIX` for Airflow. It was introduced as part of a very specific customer-driven request for Step Functions which may not apply to Airflow.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+def make_flow(
+    obj,
+    dag_name,
+    tags,
+    namespace,
+    max_workers,
+    is_project,
+    file_path=None,
+    worker_pool=None,
+    set_active=False,
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py:73]: What does `set_active` do here?

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 11]
```diff
@@ -0,0 +1,606 @@
...
+            "--monitor=%s" % self.monitor.monitor_type,
+            "--no-pylint",
+            "--with=airflow_internal",
+        ]
+
+        if node.name == "start":
+            # We need a separate unique ID for the special _parameters task
+            task_id_params = "%s-params" % self.task_id
+            # Export user-defined parameters into runtime environment
+            param_file = "".join(
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py:505]: Is this the only approach for handling parameters in Airflow? Don't they have first-class support for them?

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 12]
```diff
@@ -0,0 +1,606 @@
...
+            exists = entrypoint + [
+                # Dump the parameters task
+                "dump",
+                "--max-value-size=0",
+                "%s/_parameters/%s" % (self.run_id, task_id_params),
+            ]
+            cmd = "if ! %s >/dev/null 2>/dev/null; then %s && %s; fi" % (
+                " ".join(exists),
+                export_params,
+                bash_capture_logs(" ".join(params)),
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py]: There shouldn't be any need for `bash_capture_logs` anymore.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 13]
```diff
@@ -0,0 +1,606 @@
...
+            # todo step.append("--split-index $METAFLOW_SPLIT_INDEX")
+            pass
+        if self.tags:
+            step.extend("--tag %s" % tag for tag in self.tags)
+        if self.namespace is not None:
+            step.append("--namespace=%s" % self.namespace)
+        cmds.append(bash_capture_logs(" ".join(entrypoint + top_level + step)))
+        return " && ".join(cmds)
+
+    def _validate_workflow(self):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py]: This check can be part of Airflow CLI itself. Currently, this check executes after you have packaged and stored the code in the datastore.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 14]
```diff
@@ -0,0 +1,606 @@
...
+
+            elif node.type == "split":
+                workflow.add_state(state)
+                for func in node.out_funcs:
+                    _visit(
+                        self.graph[func],
+                        workflow,
+                    )
+
+            elif node.type == "foreach":
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py]: Since foreaches don't work right now, we can remove this conditional so that the next conditional can throw a pretty error.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+        username=get_username(),
+        is_project=is_project,
+        description=obj.flow.__doc__,
+        file_path=file_path,
+        set_active=set_active,
+    )
+
+
+@airflow.command(help="Create an airflow workflow from this metaflow workflow")
+@click.argument("file_path", required=False)
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: Let's make this a required argument

File: metaflow/plugins/airflow/airflow_cli.py [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+)
+@click.option(
+    "--name",
+    default=None,
+    type=str,
+    help="`dag_id` of airflow DAG. The flow name is used instead "
+    "if this option is not specified",
+)
+@click.option(
+    "--set-active",
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: let's name this to `--is-paused-upon-creation` so that it's more intuitive to folks coming from the Airflow world

File: metaflow/plugins/airflow/airflow_cli.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+    max_workers=None,
+    worker_pool=None,
+):
+    flow = make_flow(
+        obj,
+        name,
+        tags,
+        user_namespace,
+        max_workers,
+        False,
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: Need to enable support for `@project`

File: metaflow/plugins/airflow/airflow_cli.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+        worker_pool=worker_pool,
+        set_active=set_active,
+    )
+    compiled_dag_file = flow.compile()
+    if file_path is None:
+        obj.echo_always(compiled_dag_file)
+    else:
+        if file_path.startswith("s3://"):
+            with S3() as s3:
+                s3.put(file_path, compiled_dag_file)
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: The S3 location for DAGs may not be the same S3 location for the datastore (and likely that the user has no way of authenticating to this bucket). After the demo - let's omit this functionality.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 19]
```diff
@@ -0,0 +1,606 @@
...
+from metaflow.util import compress_list, dict_to_cli_options, to_pascalcase
+from metaflow.plugins.timeout_decorator import get_run_time_limit_for_task
+import os
+from metaflow.mflog import bash_capture_logs
+import random
+import string
+import json
+from metaflow.decorators import flow_decorators
+from metaflow.graph import FlowGraph, DAGNode
+from metaflow.plugins.cards.card_modules import chevron
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py:26]: Let's move `chevron` to `_vendor` - you make a new independent PR on top of master.

File: metaflow/plugins/airflow/airflow_compiler.py [hunk 20]
```diff
@@ -0,0 +1,606 @@
...
+from .airflow_utils import (
+    TASK_ID_XCOM_KEY,
+    Workflow,
+    AirflowTask,
+    AirflowDAGArgs,
+    AIRFLOW_TASK_ID_TEMPLATE_VALUE,
+)
+from . import airflow_utils as af_utils
+from .compute.k8s import create_k8s_args
+import metaflow.util as util
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_compiler.py]: Minor nit - let's run `isort` on the imports before merging.

File: metaflow/plugins/airflow/airflow_compiler.py (hunk 19 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_compiler.py:26]: This is non-trivial as a lot of outside "metaflow" code uses chevron. An example is outside cards we / others may have published.

File: metaflow/plugins/airflow/airflow_cli.py (hunk 10 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_cli.py:73]: `set_active=True` means `is_paused_upon_creation=False`

File: metaflow/plugins/airflow/airflow_compiler.py (hunk 12 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_compiler.py]: Why so ?

File: metaflow/datastore/task_datastore.py (hunk 7 above)
Comment (valayDave) [reply metaflow/datastore/task_datastore.py]: Yes yes. This change was for something else but i think it got bundled along mistakenly.

File: metaflow/plugins/airflow/README.md [hunk 21]
```diff
@@ -0,0 +1,19 @@
+# Airflow Integration
```
Comment (savingoyal) [inline metaflow/plugins/airflow/README.md]: We can remove this README for now

File: metaflow/plugins/airflow/af_deploy.py [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import base64
```
Comment (savingoyal) [inline metaflow/plugins/airflow/af_deploy.py]: The import statement should be inside `{{AIRFLOW_UTILS}}`

File: metaflow/plugins/airflow/airflow_cli.py [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+            return AIRFLOW_STATE_MACHINE_PREFIX + "_" + name
+        return name
+
+    project = current.get("project_name")
+    if project:
+        if name:
+            raise MetaflowException(
+                "--name is not supported for @projects. " "Use --branch instead."
+            )
+        state_machine_name = attach_prefix(current.project_flow_name)
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: `state_machine` isn't a concept in Airflow. Let's replace it with `dag_name`.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def resolve_state_machine_name(name):
+    def attach_prefix(name):
+        if AIRFLOW_STATE_MACHINE_PREFIX is not None:
+            return AIRFLOW_STATE_MACHINE_PREFIX + "_" + name
+        return name
+
+    project = current.get("project_name")
+    if project:
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py:39]: Can you verify that this logic is consistent with the implementation in `argo-workflows` and `step-functions`? Are there any limitations to the length of the DAG name in Airflow?

File: metaflow/plugins/airflow/airflow_cli.py (hunk 10 above)
Comment (savingoyal) [reply metaflow/plugins/airflow/airflow_cli.py:73]: Let's replace it with `--is-paused-upon-creation` so that it is more intuitive to folks coming from the Airflow universe.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+from .airflow_compiler import Airflow
+from metaflow import S3
+from metaflow import current
+from metaflow.exception import MetaflowException
+import re
+
+VALID_NAME = re.compile("[^a-zA-Z0-9_\-\.]")
+
+
+def resolve_state_machine_name(name):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: `resolve_dag_name`?

File: metaflow/plugins/airflow/airflow_cli.py [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+    project = current.get("project_name")
+    if project:
+        if name:
+            raise MetaflowException(
+                "--name is not supported for @projects. " "Use --branch instead."
+            )
+        state_machine_name = attach_prefix(current.project_flow_name)
+        is_project = True
+    else:
+        if name and VALID_NAME.search(name):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py:46]: Does this hold true for Airflow?

File: metaflow/plugins/airflow/airflow_cli.py [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+    "--max-workers",
+    default=100,
+    show_default=True,
+    help="Maximum number of concurrent airflow tasks to run for the DAG. ",
+)
+@click.option(
+    "--worker-pool",
+    default=None,
+    show_default=True,
+    help="Worker pool the for the airflow tasks.",
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py:153]: Nit: grammar.

File: metaflow/plugins/airflow/airflow_cli.py [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+@click.option(
+    "--namespace",
+    "user_namespace",
+    default=None,
+)
+@click.option(
+    "--max-workers",
+    default=100,
+    show_default=True,
+    help="Maximum number of concurrent airflow tasks to run for the DAG. ",
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_cli.py]: Just `Maximum number of concurrent Airflow tasks.`

File: metaflow/plugins/airflow/airflow_decorator.py [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+        run_id,
+        task_id,
+        flow,
+        graph,
+        retry_count,
+        max_user_code_retries,
+        ubf_context,
+        inputs,
+    ):
+        # todo : find out where the execution is taking place.
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_decorator.py]: Can you expand on this TODO item?

File: metaflow/plugins/airflow/airflow_decorator.py [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+        # Once figured where the execution is happening then we can do
+        # handle xcom push / pull differently
+        meta = {}
+        meta["airflow-execution"] = os.environ["METAFLOW_RUN_ID"]
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_decorator.py]: What is `airflow-execution`? We already have the `METAFLOW_RUN_ID` by virtue of the fact that you need one to be able to access this metadata.

File: metaflow/plugins/airflow/airflow_decorator.py [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+        ]
+        # Register book-keeping metadata for debugging.
+        metadata.register_metadata(run_id, step_name, task_id, entries)
+        push_xcom_values(
+            {
+                TASK_ID_XCOM_KEY: os.environ["METAFLOW_AIRFLOW_TASK_ID"],
+            }
+        )
+
+    def task_finished(
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_decorator.py]: We can omit overriding this function for now.

File: metaflow/plugins/airflow/compute/k8s.py [hunk 32]
```diff
@@ -0,0 +1,195 @@
+from datetime import timedelta
```
Comment (savingoyal) [inline metaflow/plugins/airflow/compute/k8s.py]: This can live entirely inside `airflow_compiler.py` given that we are only assembling a simple dictionary here.

File: metaflow/plugins/airflow/exceptions.py [hunk 33]
```diff
@@ -0,0 +1,20 @@
+from metaflow.exception import MetaflowException
+
+
+class AirflowNotPresent(MetaflowException):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/exceptions.py]: Where is this exception thrown?

File: metaflow/plugins/airflow/exceptions.py [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class AirflowException(MetaflowException):
+    headline = "Airflow Exception"
+
+    def __init__(self, msg):
+        super().__init__(msg)
+
+
+class NotSupportedException(MetaflowException):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/exceptions.py]: This exception doesn't seem to be used right now anywhere.

File: metaflow/plugins/airflow/exceptions.py [hunk 35]
```diff
@@ -0,0 +1,20 @@
...
+
+class AirflowNotPresent(MetaflowException):
+    headline = "Airflow cannot be imported"
+
+    def __init__(self):
+        msg = "Airflow dependency is missing. Please install airflow to use the airflow cli command"
+        super().__init__(msg)
+
+
+class AirflowException(MetaflowException):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/exceptions.py]: We can keep this exception in `airflow_compiler.py` and reduce file sprawl.

File: metaflow/plugins/airflow/plumbing/airflow_xcom_push.py [hunk 36]
```diff
@@ -0,0 +1,18 @@
+# How does xcom push work in K8s Operator : https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html#how-does-xcom-work
+import os
```
Comment (savingoyal) [inline metaflow/plugins/airflow/plumbing/airflow_xcom_push.py]: Minor - this module can simply reside in the parent directory

File: metaflow/plugins/airflow/plumbing/airflow_xcom_push.py [hunk 37]
```diff
@@ -0,0 +1,18 @@
...
+import json
+
+K8S_XCOM_DIR_PATH = "/airflow/xcom"
+
+
+def safe_mkdir(dir):
+    try:
+        os.makedirs(dir)
```
Comment (savingoyal) [inline metaflow/plugins/airflow/plumbing/airflow_xcom_push.py]: What do we need to do to ensure that this directory is always writable? Can we throw a pretty error to the user if we detect that this isn't the case? For `argo-workflows`, we get around this limitation by mounting an emptyDir volume but it seems that isn't an option in Airflow.

File: metaflow/plugins/airflow/plumbing/airflow_xcom_push.py (hunk 36 above)
Comment (savingoyal) [reply metaflow/plugins/airflow/plumbing/airflow_xcom_push.py]: Actually - this can wholly reside in `airflow_decorator.py`.

File: metaflow/plugins/airflow/plumbing/set_parameters.py [hunk 38]
```diff
@@ -0,0 +1,24 @@
+import os
+import json
+import sys
+
+
+def export_parameters(output_file):
+    input = json.loads(os.environ.get("METAFLOW_PARAMETERS", "{}"))
```
Comment (savingoyal) [inline metaflow/plugins/airflow/plumbing/set_parameters.py:7]: Do we need to handle parameters this way for Airflow? This specific approach was a workaround for all the limitations Step Functions was imposing on us. Perhaps there is a cleaner, more Airflow-native way of passing parameters?

File: metaflow/plugins/airflow/airflow_utils.py [hunk 39]
```diff
@@ -0,0 +1,477 @@
...
+# AIRFLOW_TASK_ID_TEMPLATE_VALUE uses a jinja filter called `task_id_creator` which helps
+# concatenate the string using a `/`. Since run-id will keep changing and stepname will be
+# the same task id will change. Since airflow doesn't encourage dynamic rewriting of dags
+# we can rename steps in a foreach with indexes (eg. `stepname-$index`) to create those steps.
+# Hence : Foreachs will require some special form of plumbing.
+# https://stackoverflow.com/questions/62962386/can-an-airflow-task-dynamically-generate-a-dag-at-runtime
+AIRFLOW_TASK_ID_TEMPLATE_VALUE = "arf-{{ [run_id, ti.task_id ] | task_id_creator  }}"
+
+
+def sanitize_label_value(val):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_utils.py:33]: JFYI - I will be removing this functionality in @kubernetes. Hashing this way has limited utility right now and it's better for us to understand the use cases rather than shipping a breaking change to the user later.

File: metaflow/plugins/airflow/airflow_utils.py [hunk 40]
```diff
@@ -0,0 +1,477 @@
...
+
+def hasher(my_value):
+    return hashlib.md5(my_value.encode("utf-8")).hexdigest()
+
+
+def task_id_creator(lst):
+    # This is a filter which creates a hash of the run_id/step_name string.
+    # Since run_ids in airflow are constants, they don't create an issue with the
+    #
+    return hashlib.md5("/".join(lst).encode("utf-8")).hexdigest()
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_utils.py]: Can you truncate the task-id to say 8 chars or so? They need to be unique within a step and it's a good guarantee that we won't run into collisions.

Comment (savingoyal) [reply metaflow/plugins/airflow/airflow_utils.py]: Similar for the run-id that is generated. Since that needs to be globally unique - we can go with 12 chars.

File: metaflow/plugins/airflow/airflow_utils.py [hunk 41]
```diff
@@ -0,0 +1,477 @@
...
+
+
+def json_dump(val):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_utils.py:63]: This function isn't really doing anything. Do we need it?

File: metaflow/plugins/airflow/airflow_utils.py [hunk 42]
```diff
@@ -0,0 +1,477 @@
...
+            "trigger_rule": "all_success"
+            # 'dag': dag,
+            # 'on_failure_callback': some_function,
+            # 'on_success_callback': some_other_function,
+            # 'on_retry_callback': another_function,
+            # 'sla_miss_callback': yet_another_function,
+        },
+    }
+
+    metaflow_centric_args = {
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_utils.py]: Can we have a better name for this?

File: metaflow/plugins/airflow/airflow_utils.py [hunk 43]
```diff
@@ -0,0 +1,477 @@
...
+    # Since run_ids in airflow are constants, they don't create an issue with the
+    #
+    return hashlib.md5("/".join(lst).encode("utf-8")).hexdigest()
+
+
+def json_dump(val):
+    return json.dumps(val)
+
+
+class AirflowDAGArgs(object):
```
Comment (savingoyal) [inline metaflow/plugins/airflow/airflow_utils.py:67]: This entire module can just live inside `af_deploy.py`?

File: metaflow/plugins/airflow/exceptions.py (hunk 34 above)
Comment (valayDave) [reply metaflow/plugins/airflow/exceptions.py]: Used when we detect foreachs.

File: metaflow/plugins/airflow/airflow_utils.py (hunk 41 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_utils.py:63]: it is used as a `user_defined_filter` for airflow DAG

File: metaflow/plugins/airflow/plumbing/airflow_xcom_push.py (hunk 37 above)
Comment (valayDave) [reply metaflow/plugins/airflow/plumbing/airflow_xcom_push.py]: Airflow communicates XCOM via a sidecar container that has `emptyDir` volume mounted to `/airflow/xcom`. There could be a condition where the side car has not started when the job container starts execution. So this condition ensures that directories are made safely. I don't think we need to raise any exception here.

File: metaflow/plugins/airflow/airflow_decorator.py (hunk 30 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_decorator.py]: Spurious. Will Remove.

File: metaflow/plugins/airflow/airflow_decorator.py (hunk 29 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_decorator.py]: XCOM's would work differently based on computing mediums. 

Example:  K8s Operators leverages XCOM via reading a JSON file present in the volume mounted by its sidecar container. 

So if we introduce Local/batch etc execution then we need to figure out how to pass the XCOM values.

File: metaflow/plugins/airflow/plumbing/set_parameters.py (hunk 38 above)
Comment (valayDave) [reply metaflow/plugins/airflow/plumbing/set_parameters.py:7]: Can you expand on what you mean by a more airflow native way? I use the `{{params}}` macro from airflow to pass the parameters to an environment variable. These are then used to set `METAFLOW_INIT_` env variable.

File: metaflow/plugins/airflow/airflow_cli.py (hunk 17 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_cli.py]: Project support is there. We switch names of the dag. I don't how else would I include it. I will clean up this bit from the code.

File: metaflow/plugins/airflow/airflow_cli.py (hunk 24 above)
Comment (valayDave) [reply metaflow/plugins/airflow/airflow_cli.py:39]: I researched on dagname length limitations and couldn't find any explicitly. Will keep looking into it.

File: metaflow/plugins/airflow/af_deploy.py (hunk 22 above)
Comment (valayDave) [reply metaflow/plugins/airflow/af_deploy.py]: Which import statement? Do you mean `import base64` ? I think It's a rogue import. I will remove it.
